scene/2d/camera_2d.cpp

```diff
@@ -33,183 +33,223 @@
-			}
-
-			if (screen_rect.position.y + screen_rect.size.y > limit[SIDE_BOTTOM]) {
-				camera_pos.y -= screen_rect.position.y + screen_rect.size.y - limit[SIDE_BOTTOM];
+		if (drag_vertical_offset_changed) {
+			if (drag_vertical_offset > 0) {
+				target_position.y += drag_vertical_offset * drag_margin[SIDE_BOTTOM] * half_screen_size.y;
+			} else { // drag_vertical_offset < 0
+				target_position.y += drag_vertical_offset * drag_margin[SIDE_TOP] * half_screen_size.y;
 			}
-
-			if (screen_rect.position.y < limit[SIDE_TOP]) {
-				camera_pos.y -= screen_rect.position.y - limit[SIDE_TOP];
+			drag_vertical_offset_changed = false;
+		} else if (drag_vertical_enabled && !Engine::get_singleton()->is_editor_hint()) {
+			if (target_position.y > current_center.y + drag_margin[SIDE_BOTTOM] * half_screen_size.y) {
+				target_position.y += drag_margin[SIDE_BOTTOM] * half_screen_size.y;
+			} else if (target_position.y < current_center.y - drag_margin[SIDE_TOP] * half_screen_size.y) {
+				target_position.y -= drag_margin[SIDE_TOP] * half_screen_size.y;
+			} else {
+				target_position.y = current_center.y;
 			}
 		}
+		target_position -= half_screen_size;
+	}
 
-		if (smoothing_enabled && !Engine::get_singleton()->is_editor_hint()) {
-			real_t c = smoothing * (process_callback == CAMERA2D_PROCESS_PHYSICS ? get_physics_process_delta_time() : get_process_delta_time());
-			smoothed_camera_pos = ((camera_pos - smoothed_camera_pos) * c) + smoothed_camera_pos;
-			ret_camera_pos = smoothed_camera_pos;
-			//camera_pos=camera_pos*(1.0-smoothing)+new_camera_pos*smoothing;
-		} else {
-			ret_camera_pos = smoothed_camera_pos = camera_pos;
-		}
-
-	} else {
-		ret_camera_pos = smoothed_camera_pos = camera_pos = new_camera_pos;
-		first = false;
+	Vector2 limit_adjust;
+	if (target_position.x < limit[SIDE_LEFT]) {
+		limit_adjust.x += limit[SIDE_LEFT] - target_position.x;
+	}
+	if (target_position.x > limit[SIDE_RIGHT] - screen_size.x) {
+		limit_adjust.x += limit[SIDE_RIGHT] - screen_size.x - target_position.x;
 	}
+	if (target_position.y < limit[SIDE_TOP]) {
+		limit_adjust.y += limit[SIDE_TOP] - target_position.y;
+	}
+	if (target_position.y > limit[SIDE_BOTTOM] - screen_size.y) {
+		limit_adjust.y += limit[SIDE_BOTTOM] - screen_size.y - target_position.y;
+	}
+	target_position += limit_adjust;
 
-	Point2 screen_offset = (anchor_mode == ANCHOR_MODE_DRAG_CENTER ? (screen_size * 0.5 * zoom_scale) : Point2());
+	if (!smoothing_active) {
+		current_position = target_position;
+		_update_viewport();
+	}
+}
 
-	real_t angle = get_global_rotation();
-	if (rotating) {
-		screen_offset = screen_offset.rotated(angle);
+// Must only be called by internal process notification!
+void Camera2D::_update_scroll() {
+	if (!initialized) {
+		_update_position();
 	}
 
-	Rect2 screen_rect(-screen_offset + ret_camera_pos, screen_size * zoom_scale);
+	if (!smoothing_active) {
+		current_position = target_position;
+		_update_viewport();
+	}
```

Comment:
Just to check, should this be passing through to the rest below, or should there be a return here?

Reply:
Yes, there should be a return here. Good spot!